clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1873,4 +1875,70 @@ void handleFailure(Throwable throwable) {
         return new CreatePartitionsResult(new HashMap<String, KafkaFuture<Void>>(futures));
     }
 
+    @Override
+    public ElectPreferredLeadersResult electPreferredLeaders(final Collection<TopicPartition> partitions,
+                                                             ElectPreferredLeadersOptions options) {
+
+        final KafkaFutureImpl<Map<TopicPartition, KafkaFutureImpl<Void>>> futures = new KafkaFutureImpl<>();
+        final Map<TopicPartition, KafkaFutureImpl<Void>> mapOfFutures;
+        final List<TopicPartition> requestList;
+        final boolean knownPartitions = partitions != null;
+        if (knownPartitions) {
+            mapOfFutures = new HashMap<>(partitions.size());
+            for (TopicPartition partition : partitions) {
+                KafkaFutureImpl future = new KafkaFutureImpl<Void>();
+                mapOfFutures.put(partition, future);
+            }
+            futures.complete(mapOfFutures);
+            requestList = new ArrayList<>(partitions);
+        } else {
+            mapOfFutures = new HashMap<>();
+            requestList = null;
+        }
+
+        final long now = time.milliseconds();
+        runnable.call(new Call("electPreferredLeaders", calcDeadlineMs(now, options.timeoutMs()),
+                new ControllerNodeProvider()) {
+
+            @Override
+            public AbstractRequest.Builder createRequest(int timeoutMs) {
+                return new ElectPreferredLeadersRequest.Builder(requestList, timeoutMs);
+            }
+
+            @Override
+            public void handleResponse(AbstractResponse abstractResponse) {
+                ElectPreferredLeadersResponse response = (ElectPreferredLeadersResponse) abstractResponse;
+                // Iterate over the response partitions->errors because the argument partitions can be null
+                for (Map.Entry<TopicPartition, ApiError> entry : response.errors().entrySet()) {
+                    TopicPartition partition = entry.getKey();
+                    KafkaFutureImpl<Void> future;
+                    if (knownPartitions) {
+                        future = mapOfFutures.get(partition);
+                    } else {
+                        future = new KafkaFutureImpl<Void>();
+                        mapOfFutures.put(partition, future);
```

Comment:
You can't modify this map here because it is concurrently accessed by people accessing ElectPreferredLeadersResult.

Reply:
@cmccabe there are two cases here, depending on the state of `knownPartitions`. We only modify `mapOfFutures` when `knownPartitions == false` and in this case it's impossible for anyone to access the map via the `ElectPreferredLeadersResult` because the "outer" `futures` won't be completed.